src/dataflow-types/src/client.rs

```diff
@@ -1143,6 +1369,177 @@ pub mod tcp {
+    /// network, but put messages in a mpsc queue which will be read and sent to the network in a
+    /// separate server task.
+    pub struct GrpcComputedServerInterface {
+        shared: Arc<GrpcComputedShared>,
+    }
+
+    impl GrpcComputedServerInterface {
+        pub async fn send(&self, resp: ComputeResponse) -> () {
+            loop {
+                match self.shared.queue.lock().await.as_ref() {
+                    Some(x) => {
+                        x.1.send((&resp).into_proto()).await.unwrap();
+                        break;
+                    }
+
+                    // Other end absent, wait for connection
+                    None => self.shared.queue_change.notified().await,
+                }
+            }
+        }
+
+        pub async fn recv(&mut self) -> Result<ComputeCommand, TryFromProtoError> {
+            loop {
+                match self.shared.queue.lock().await.as_mut() {
+                    Some(x) => match x.0.recv().await {
+                        Some(x) => return x.into_rust(),
+                        None => {
+                            println!("Recv returned None, endpoint gone, retry");
+                        }
+                    },
+
+                    // Other end absent, wait for connection
+                    None => {
+                        info!("recv called while no coordinator connected, waiting for coordinator connection.");
+                    }
+                }
+
+                self.shared.queue_change.notified().await;
+            }
+        }
+    }
+
+    type ResponseStream =
+        Pin<Box<dyn Stream<Item = Result<super::ProtoComputeResponse, tonic::Status>> + Send>>;
+
+    #[tonic::async_trait]
+    impl super::proto_compute_server::ProtoCompute for GrpcComputedServer {
+        type HandleStream = ResponseStream;
+
+        async fn handle(
+            &self,
+            req: Request<Streaming<super::ProtoComputeCommand>>,
+        ) -> Result<tonic::Response<Self::HandleStream>, tonic::Status> {
+            info!("GrpcComputedServer: remote client connected");
+
+            let mut in_stream = req.into_inner();
+            // The channel sizes are arbitrary, but it should be a
+            // small bounded channel such that backpressure is applied early.
+            let (cmd_tx, cmd_rx) = mpsc::channel(4);
+            let (resp_tx, resp_rx) = mpsc::channel(4);
+
+            // this spawn here is required if you want to handle connection error.
+            // If we just map `in_stream` and write it back as `out_stream` the `out_stream`
+            // will be drooped when connection error occurs and error will never be propagated
+            // to mapped version of `in_stream`.
+            let linger = self.linger;
+            let shutdown_tx = self.shutdown_tx.clone();
+            mz_ore::task::spawn(|| "ProtoComputeServer: ErrorChecker", async move {
```

Comment:
is this the correct name? this is the task that actually pull values off the input stream

separately, do we need `cmd_tx/cmd_rx`? why not just put the `in_stream` in the `shared.queue` and just call `.next()` on it in `recv`?